core/src/banking_stage.rs

```diff
@@ -1060,6 +1078,11 @@ impl BankingStage {
                 let tx: Transaction = limited_deserialize(&p.data[0..p.meta.size]).ok()?;
                 tx.verify_precompiles(libsecp256k1_0_5_upgrade_enabled)
                     .ok()?;
+                vote_redundancy_checker
+                    .write()
+                    .unwrap()
+                    .check(&tx, bank)
+                    .ok()?;
```

Comment:
Hmm I think there's a race here. Even if the `vote_redundancy_checker` see slots `1->2->3->4` in order, the execution could still be out of order because ultimately what matters is the order in which these banking threads hit the `process_and_record_transactions_locked()` and grab the locks for the accounts

Reply:
on second thought, if banking_stage tracks last voted slot, and assume all transaction it let through will eventually landed in blockstore, it should be able to drop votes that voted for older slots (as implemented as solution 2 at https://github.com/solana-labs/solana/pull/19152/files#diff-bb8464c2f02c8863fb60e6e29f023f3e6da93a93fbb9199a70a44c9c1b41c7f3R1103-R1118 ). Do you think so?

Reply:
Hmm, this still seems to run into the same issue. I can have two threads, one gets a vote for slot 1, one for slot 2 from the same validator. They insert the slots in order into `latest_voted_slots`, but the thread with slot 2 processes it's vote transaction first. Then the thread adding `slot 1` will incur the `TooOld` error when it tries to be process the transaction.

Reply:
right! the "solution 1", which banking_stage checks vote without involving `bank`, inherently has chances to pass old votes to bank for processing, though the possibilities are largely reduced (by filtering out out-of-order votes received by banking_stage). The benefit of this approach, as mentioned in Summary of Change above, is it doesn't require `bank`, which alights with bank-less leader goal. 

The other solution implemented in this PR (https://github.com/solana-labs/solana/pull/19152/files#diff-ed47b4a0198313377e091bb3957bbbc63d937805426d1b2b6de39d0a50d32a0cR3461) checks vote in bank without race condition, but my concern is if this approach will be invalidated when implementing bank-less leader (I haven't thought it through, just a general concern) 

At the end, we only need one of these two implementations in this PR. The debate is "which one" :)

Reply:
A few things:

1) Doesn't the bankless leader design also require access to a bank in order to figure out whether an account has sufficient funds to pay the transaction fee?

2) It seems like the stateless `VoteRedundancyChecker` could theoretically also be used in solution 2 *after* we grab the locks? 

3) If we go with the `VoteRedundancyChecker`,  I also think we may need a separate `VoteRedundancyChecker` per bank because of forking. For example if you had

```
      1
  /       \
2          3
```

A vote for slot 1 may have landed in the bank for slot 2, but not for the bank in slot 3.

Reply:
1. Yeah, will need bank to check account balance, would be great is just simply in-n-out for that purpose only. But I am not complete sure, not opposing to do check after account lock at all. 
2. Yep, the Bank solution currently does the same thing `VoteRedundantChecker` does. 
3. This is another point for using bank-based solution, so current implement (https://github.com/solana-labs/solana/pull/19152/files#diff-ed47b4a0198313377e091bb3957bbbc63d937805426d1b2b6de39d0a50d32a0cR3117) covers the forking scenario, i think

Answering though these questions, I started to see the benefit of bank-based solution, I'll remove the banking_stage based code, to make the PR easier for review.